src/blueprint.zig

```diff
@@ -0,0 +1,281 @@
+const std = @import("std");
+
+const main = @import("main.zig");
+const Compression = main.utils.Compression;
+const ZonElement = @import("zon.zig").ZonElement;
+const vec = main.vec;
+const Vec3i = vec.Vec3i;
+
+const Block = main.blocks.Block;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const User = main.server.User;
+
+pub const blueprintVersion = 0;
+pub const GameIdToBlueprintIdMapType = std.AutoHashMap(u16, u16);
+const BlockIdSizeType = u32;
+const BlockStorageType = u32;
+
+const BinaryWriter = main.utils.BinaryWriter;
+const BinaryReader = main.utils.BinaryReader;
+
+pub const BlueprintCompression = enum(u16) {
+	deflate,
+};
+
+pub const Blueprint = struct {
+	blocks: main.List(Block),
+	sizeX: usize,
+	sizeY: usize,
+	sizeZ: usize,
+
+	pub fn init(allocator: NeverFailingAllocator) @This() {
+		return Blueprint{
+			.blocks = .init(allocator),
+			.sizeX = 0,
+			.sizeY = 0,
+			.sizeZ = 0,
+		};
+	}
+	pub fn deinit(self: *@This(), _: NeverFailingAllocator) void {
+		self.blocks.deinit();
+	}
+	pub fn clear(self: *@This()) void {
+		self.sizeX = 0;
+		self.sizeY = 0;
+		self.sizeZ = 0;
+		self.blocks.clearRetainingCapacity();
+	}
+	pub fn clone(self: *@This(), allocator: NeverFailingAllocator) @This() {
+		var new = Blueprint.init(allocator);
+		new.sizeX = self.sizeX;
+		new.sizeY = self.sizeY;
+		new.sizeZ = self.sizeZ;
+		new.blocks = .initCapacity(allocator, self.blocks.items.len);
+		for(self.blocks.items) |block| {
+			new.blocks.appendAssumeCapacity(block);
+		}
+		return new;
+	}
+	pub fn capture(self: *@This(), pos1: Vec3i, pos2: Vec3i) ?struct {x: i32, y: i32, z: i32, message: []const u8} {
+		self.clear();
+
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: usize = @abs(endX - startX + 1);
+		const sizeY: usize = @abs(endY - startY + 1);
+		const sizeZ: usize = @abs(endZ - startZ + 1);
```

Comment:
Should use `@intCast` instead of `@abs`, since we know that end is always larger than start.